modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java

```diff
@@ -81,4 +86,68 @@ public static int findStartBuildingIndexCatalogVersion(CatalogService catalogSer
                 BUILDING, indexId, fromCatalogVersionIncluded, latestCatalogVersion
         ));
     }
+
+    /**
+     * Collects a list of tables that were removed from the catalog and should have been dropped due to a low watermark (if the catalog
+     * version in which the table was removed is less than or equal to the active catalog version of the low watermark).
+     *
+     * @param catalogService Catalog service.
+     * @param lowWatermark Low watermark, {@code null} if it has never been updated.
+     */
+    // TODO: IGNITE-21771 Process or check catalog compaction
+    static List<DroppedTableInfo> droppedTables(CatalogService catalogService, @Nullable HybridTimestamp lowWatermark) {
+        if (lowWatermark == null) {
+            return List.of();
+        }
+
+        int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+        int lwmCatalogVersion = catalogService.activeCatalogVersion(lowWatermark.longValue());
+
+        var tableIds = catalogService.tables(lwmCatalogVersion).stream()
+                .map(CatalogObjectDescriptor::id)
+                .collect(toSet());
+
+        var res = new ArrayList<DroppedTableInfo>();
+
+        for (int catalogVersion = lwmCatalogVersion - 1; catalogVersion >= earliestCatalogVersion; catalogVersion--) {
+            int finalCatalogVersion = catalogVersion;
+
+            catalogService.tables(catalogVersion).stream()
+                    .map(CatalogObjectDescriptor::id)
+                    .filter(tableIds::add)
+                    .map(tableId -> new DroppedTableInfo(tableId, finalCatalogVersion + 1))
+                    .forEach(res::add);
+        }
+
+        return res;
+    }
+
+    static List<DroppedTableInfo> droppedTables0(CatalogService catalogService, @Nullable HybridTimestamp lowWatermark) {
```

Comment:
Looks like you forgot to remove the previous version

Reply:
fix it